File: easy/paranthesis/Go/paranthesis.go
package main

// Cathegory: Easy
// Success
// Details
// Runtime: 3 ms, faster than 35.99% of Go online submissions for Valid Parentheses.
// Memory Usage: 2 MB, less than 43.19% of Go online submissions for Valid Parentheses.

// 92 / 92 test cases passed.
// 	Status: Accepted
// Runtime: 3 ms
// Memory Usage: 2 MB

// copyright 2022 Bogdan Peta 

import (
	// "fmt"
)

func validParant(s string) bool {

	temp := []rune{}
	pairs := map[rune]rune{'(': ')', '[': ']', '{': '}'}

	for _, ch := range s {
		if len(temp) == 0 {
			temp = append(temp, pairs[ch])
		} else {
			last := temp[len(temp) - 1]
			if ch == last {
				temp = temp[0:len(temp) - 1]
			} else {
				temp = append(temp, pairs[ch])
			}
		}
	}
	sz := len(temp);

	return sz == 0
}

File: easy/paranthesis/Rust/src/paranthesis.rs
// Success
// Details
// Runtime: 2 ms, faster than 43.15% of Rust online submissions (2,800,170) for Valid Parentheses.
// Memory Usage: 2.2 MB, less than 40.88% of Rust online submissions (2,800,170) for Valid Parentheses.

// 92 / 92 test cases passed.
// 	Status: Accepted
// Runtime: 2 ms
// Memory Usage: 2.2 MB

// Date: 7.11.2022

// Copyright 2022 Bogdan Peta, All Rights Reserved

use std::collections::HashMap;

// In Go if a key does not exists in map map[key] 
// returns nil for that type, 0 for integer in our case
// Rust panics when a key does not exists, so emulate 
// that with unwrap_or


pub fn match_paranthesis(s: String) -> bool
{
    if s.len() < 1 || s.len() > 10000 {
        return false;
    }

    let mut temp: Vec<u8> = Vec::new();

    let pairs: HashMap<u8, u8> = [
        (40,41), (91,93), (123,125),
    ].into();

    for ch in s.bytes() {
        if temp.len() == 0 {
            temp.push(*pairs.get(&ch).unwrap_or(&0));
        } else {    
            let last = temp[temp.len() - 1];
            if ch == last {
                temp.pop();
            } else {
                temp.push(*pairs.get(&ch).unwrap_or(&0));
            }
        }
    }

    temp.len() == 0
}
